<p style="font-size: 25px; text-align: center; color: #722F37	;"><strong>Tunisian Video Comment Extraction for Sentiment Analysis</strong></p>

<font color='AA336A' style="font-size: 20px;"><strong>Table of contents:</strong><font><a class='anchor' id='top'></a>
<ul>
    <li><a href= "#intro"> Introduction</a></li>
    <li><a href="#Packages" >Packages import  </a></li>
    <li><a href="#Business" >Business & Data Understanding </a></li>
    <li><a href="#Data"> Data Preapration </a></li>
    <li><a href="#Machine"> Machine Learning </a></li>
    <li><a href="#Performance"> Performance Evaluation  </a> </li>
</ul>

<font color='#AA336A' style="font-size: 20px;"><strong>1- Introduction: </strong></font>
    <p style="color: #AA98A9; font-family: Arial; font-size: 15px;"><strong> >>  Why Sentiment Analysis of YouTube Comments Matters ?</strong></p>
In the sea of data within YouTube comments, sentiment analysis emerges as a critical tool for your machine learning project. Understanding the emotional undercurrent – positive, negative, or neutral – expressed in comments unlocks a treasure trove of insights:
<p  style="font-family: Arial"><strong><u> 1) Community Segmentation and Understanding: </u></strong></p>
Analyzing comment data with ML can help you understand different segments within the Tunisian online community. This could involve identifying demographics, interests, and potential influencers based on comment patterns.<br>
<p style="font-family: Arial"><strong><u> 2) Identifying Emerging Trends:</u></strong></p>
Shifts in sentiment can reveal emerging trends and evolving public discourse surrounding a particular topic. For example, a sudden spike in negative comments following a government decision could indicate growing public dissatisfaction.
<p style="font-family: Arial"><strong><u> 3) Measuring Brand Perception:</u></strong></p>
Businesses can analyze sentiment towards their YouTube videos or those of competitors. Positive sentiment reflects brand favorability, while negative sentiment highlights areas for improvement.<br>
<p style="font-family: Arial"><strong> *** This allows companies to refine their marketing strategies and messaging. ***</strong></p>

<p style="font-family: Arial;box-sizing: border-box;
width: 100%; text-align: center;
border: solid #AA336A 3px;
padding: 5px;"><em>Being aware of this we choose to work in this context. This project is fueled by a rich database capturing the voices of the Tunisian online community. 
There is a large collection of YouTube comments, all labeled with their sentiment.
– "neutral," "positive," or "negative."Through his extensive dataset, we can explore the depths of public opinion, analyze trends, and gain valuable insights into how Tunisians feel about various topics.</em></p>





<a name="Packages"><font color='#AA336A' style="font-size: 20px;"><strong> 2- Packages import </strong></font></a>

In [1]:
import re
import nltk
import pandas as pd 
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neural_network  import MLPClassifier
from sklearn.model_selection import train_test_split 
from sklearn.metrics import classification_report 
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from nltk.corpus import stopwords 
import pickle

<a name="Business"><font color='#AA336A' style="font-size: 20px;"><strong>  3- Business & Data Understanding </strong></font></a>

In [2]:
df=pd.read_excel("dataset_final.xlsx")
df.head()

Comment    Label
0        انتم لاهين كان بالقبل المتبادلة ارتقوا شوية  negatif
1  Wech hed etmenyik ya twansa vraiment rakom thayro  negatif
2                                             خماج😬😬  negatif
3                     لمة جهنم وبئس المصير اللطف كهو  negatif
4                                               خماج  negatif

<p style="font-family: Arial;box-sizing: border-box;
width: 50%; text-align: center;
border: solid #AA336A 3px;
padding: 5px;margin: 0 auto">This database contains examples of comments mainly in <strong>Arabic</strong>, but sometimes in French. Comments can <strong>also contain emojis and numbers</strong>. In next section we further explore the characteristics of this database. </p>

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9985 entries, 0 to 9984
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Comment  9980 non-null   object
 1   Label    9984 non-null   object
dtypes: object(2)
memory usage: 156.1+ KB


<div  style="font-family: Arial;box-sizing: border-box;
width: 100%;
border: solid #AA336A 3px; padding: 5px; margin: 0 auto;"><div style="margin-left: 10px ;margin-top: 15px;"><p style="font-size: 15px;"font color='#AA336A'><em><strong>This database contains:  :</strong></em></p><br></div>
    <div style="margin-left: 20px; margin-bottom: 20px">
    <u>*** Comment *** :</u> <br>
    -  9985 lines and 2 columns of which the total number of column comments is 9980.  <br>
    -  5 comments are missing (9985 entries in total - 9980 entries with comments = 5 entries missing).<br>
    -  The Comment column data type "object" suggests that comments are stored as text strings.<br>
    </div>
<div style="margin-left: 20px;">    
<u>*** Label *** :</u> This column contains the labels associated with each comment there are:: <br>
   - 9984 Label non-zero, which indicates that there might be 1 label missing (9985 entries total - 9984 entries with labels = 1 entry missing). <br>
   - As with the comments column, the "object" data type suggests that labels are stored as text strings.</div></div>


In [4]:
df_result=df.dropna()
df_result.info()

<class 'pandas.core.frame.DataFrame'>
Index: 9980 entries, 0 to 9984
Data columns (total 2 columns):
 #   Column   Non-Null Count  Dtype 
---  ------   --------------  ----- 
 0   Comment  9980 non-null   object
 1   Label    9980 non-null   object
dtypes: object(2)
memory usage: 233.9+ KB


<div  style="font-family: Arial;box-sizing: border-box;
width: 50%;
border: solid #AA336A 3px;
             padding: 5px; margin: 0 auto;"><div style="margin-left: 10px ;margin-top: 15px;"><p style="font-size: 15px;"font color='#AA336A'><em><strong>Compared to the other database :</strong></em></p><br></div>
    <div style="margin-left: 20px; margin-bottom: 20px">
    - Deletion of 5 comments are missing (to become 9980 entries in total) <br>
    - Deletion of a missing label (to become 9980 entries in total). <br>
</div>
</div>

In [5]:
#classification mutticlass non biaire
df.Label.value_counts()

Label
negatif     4927
positive    4679
neutre       378
Name: count, dtype: int64

<div  style="font-family: Arial;box-sizing: border-box;
width: 100%;
border: solid #AA336A 3px; padding: 5px; margin: 0 auto;"><div style="margin-left: 30px ;margin-top: 15px;"><p style="font-size: 15px;"font color='#AA336A'>
    <strong>This database contains:</strong><br>
    <div style="margin-left: 20px; ">
     * 9985 comments <br>
     * 3 classes which are negative, neutral, positive <br> 
        </div>
    <div style="margin-top: 15px;">
    <strong>Data Imbalance:</strong><br>
    <div style=" margin-top: 15px;">
    - More positive and negative comments compared to neutral ones. To address this imbalance, we considered two approaches:

    1. **Technique to Make Dataset Balanced:** 
       - Oversampling (duplicate minority class examples)
       - Undersampling (remove majority class examples)
       - SMOTE (synthetic minority oversampling technique)

    2. **Algorithm Adapted to Balanced Datasets:** 
       - Support Vector Machines (SVMs) with class weighting
       - Random Forest
       - Gaussien naive bayes

    **We chose approach 2** for its efficiency and reduced risk of introducing bias through data manipulation.
</div> </div>


 <a name="Data"><font color='#AA336A' style="font-size: 20px;"><strong>4- Data preparation </strong></font></a>

<font color='#A95C68' style="font-size: 15px;"><strong>  4.1 Data cleaning  </strong></font>

In [6]:
features=df.iloc[:,0]
labels=df.iloc[:,1]

In [7]:
emoji_pattern = re.compile("["
                           u"\U0001F600-\U0001F64F"  # emoticons
                           u"\U0001F300-\U0001F5FF"  # symbols & pictographs
                           u"\U0001F680-\U0001F6FF"  # transport & map symbols
                           u"\U0001F1E0-\U0001F1FF"  # flags (iOS)
                           u"\U00002702-\U000027B0" # Miscellaneous symbols (checkmarks, musical notes, etc.)
                           u"\U000024C2-\U0001F251" # More symbols and pictographs (including arrows, geometric shapes, etc.)
                           "]+", flags=re.UNICODE) #This optional flag specifies that the regular expression engine should interpret the pattern using Unicode characters. 
regex = r"[^\u0600-\u06FF]"
tidy_features=[]
tidy_labels=[]
for i in range(len(features)):
    text=emoji_pattern.sub(r'', str(features[i])) # to remove emoji
    text=re.sub(r'https?://\S+|www\.\S+', ' ', text) # to remove link 
    text=re.sub(r"<.*?>", " ", text) #to remove html 
    text=re.sub(r'\d+', ' ', text) # to remove number
    text=re.sub(regex, " ", text) # to identify characters that are not Arabic.
    if text.strip():  # Check if text is not empty after stripping whitespace
        tidy_features.append(text)
        tidy_labels.append(labels[i])


df_cleaned = pd.DataFrame(data={'Cleaned_Text': tidy_features, 'labels': tidy_labels})
# to verify that uncessary charchter are removed from text 
df_cleaned.to_excel("cleaned_features.xlsx", index=False)

print("Preprocessing complete! Results saved to cleaned_features.xlsx")




Preprocessing complete! Results saved to cleaned_features.xlsx


In [8]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5842 entries, 0 to 5841
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Cleaned_Text  5842 non-null   object
 1   labels        5842 non-null   object
dtypes: object(2)
memory usage: 91.4+ KB


In [9]:
# After cleaning the data by removing unnecessary characters, we identified a column containing null values so we need it
# to remove it 
df_result=df_cleaned.dropna()
df_result.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5842 entries, 0 to 5841
Data columns (total 2 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   Cleaned_Text  5842 non-null   object
 1   labels        5842 non-null   object
dtypes: object(2)
memory usage: 91.4+ KB


<font color='#A95C68' style="font-size: 15px;"> <strong>4.2 Data Normalisation  </strong></font>

In [10]:
from pyarabic import araby
unified_text=[]
for i in range(len(tidy_features)):
    text = re.sub("[إأآا]", "ا", str(tidy_features[i]))
    text = re.sub("ى", "ي", text)
    text = re.sub("ؤ", "ء", text)
    text = re.sub("ئ", "ء", text)
    text = re.sub("گ", "ك", text)
    # to remove ("ضمة كسرة فتحة " "إشكال ") generally 
    text = araby.strip_diacritics(text)
    text = araby.strip_tatweel(text)
    unified_text.append(text)

# Create DataFrame and save to Excel
normalised_df = pd.DataFrame(data={'normalised_Text': unified_text, 'labels': tidy_labels})
normalised_df.to_excel("normalised.xlsx", index=False)



print("Preprocessing complete! Results saved to normalised.xlsx")


Preprocessing complete! Results saved to normalised.xlsx


<div  style="font-family: Arial;box-sizing: border-box;
width: 75%;
border: solid #AA336A 3px;padding: 5px;margin: 0 auto;"><div style="margin-left: 10px ;margin-top: 15px;"><p style="font-size: 15px;"font color='#AA336A'><em><strong>fit :</strong></em> analyzes the text data in unified_text to learn the vocabulary (unique words) and calculate term frequencies (how often each word appears).</em></div>
<div style="margin-left: 10px ;;margin-top: 10px;"><p style="font-size: 15px;"font color='#AA336A'>
    <em> <strong>Transform : </strong></em> It converts each text document in unified_text into a TF-IDF vector representation. TF-IDF considers both the word frequency within a document and its overall frequency across the entire corpus, giving more weight to words that are important within a document but less frequent overall.</div></div>

In [11]:
#all_stemmed_words = [word for sublist in stemmed_tokens for word in sublist]

vectorizer = TfidfVectorizer(stop_words=stopwords.words('arabic'))

X = vectorizer.fit_transform(unified_text).toarray()


C:\Users\Imen\anaconda3\Lib\site-packages\sklearn\feature_extraction\text.py:408: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['آمين', 'أب', 'أخ', 'أفعل', 'أفعله', 'ؤلاء', 'إل', 'إم', 'ات', 'اتان', 'ارتد', 'ان', 'انفك', 'برح', 'تان', 'تبد', 'تحو', 'تعل', 'حد', 'حم', 'حي', 'خب', 'ذار', 'سيما', 'صه', 'ظل', 'ظن', 'عد', 'قط', 'مر', 'مكان', 'مكانكن', 'نب', 'هات', 'هب', 'واها', 'وراء'] not in stop_words.
  warnings.warn(


<font color='#A95C68' style="font-size: 15px;"> <strong> 4.3 Word emedding usinf tf-idf Vectoriser </strong></font>

In [12]:
X_train,X_test,y_train,y_test=train_test_split(X,tidy_labels,test_size=0.2)

<a name="Machine"><font color='#AA336A' style="font-size: 20px;"><strong> 5- Machine learning </strong></font></a>

In [13]:
random_state = 40

In [14]:

# the use of this MLP algorithms just to show that have bad performance in our case 
neural= MLPClassifier(hidden_layer_sizes=(100,150),activation='logistic',solver='adam',random_state=random_state)
gnb= GaussianNB()
#This strategy aims to compensate for the class imbalance during SVM training
linear_svm=SVC(kernel='linear',class_weight='balanced',random_state=random_state)
rbf_svm=SVC(kernel='rbf',class_weight='balanced',random_state=random_state)
sgd_svm=SVC(kernel='sigmoid',class_weight='balanced',random_state=random_state)
poly_svm=SVC(kernel='poly',degree=2,class_weight='balanced',random_state=random_state)

#training Ml algorithms
neural.fit(X_train,y_train)
gnb.fit(X_train,y_train)
linear_svm.fit(X_train,y_train)
rbf_svm.fit(X_train,y_train)
sgd_svm.fit(X_train,y_train)
poly_svm.fit(X_train,y_train)







SVC(class_weight='balanced', degree=2, kernel='poly', random_state=40)

In [15]:
#Predection
y_pred_neural=neural.predict(X_test)
y_pred_linear=linear_svm.predict(X_test)
y_pred_rbf=rbf_svm.predict(X_test)
y_pred_sgd=sgd_svm.predict(X_test)
y_pred_poly=poly_svm.predict(X_test)
y_pred_gnd=gnb.predict(X_test)

In [16]:
from sklearn.ensemble import RandomForestClassifier


lass_weight = 'balanced'  
rf_model = RandomForestClassifier(n_estimators=100, random_state=random_state, class_weight=lass_weight,criterion='entropy')

#training Ml algorithms
rf_model.fit(X_train, y_train)


RandomForestClassifier(class_weight='balanced', criterion='entropy',
                       random_state=40)

In [17]:
#Predection
y_pred_forest = rf_model.predict(X_test)


<a name="Performance"><font color='#AA336A' style="font-size: 20px;"><strong> 6- Performance evaluation  </strong></font></a>

In [18]:
print("************* performance evaluation of Naive bayes ***********")
print(classification_report(y_test,y_pred_linear))
print("************* performance evaluation of MLP Classifier  ***********")
print(classification_report(y_test,y_pred_neural))
print("************* performance evaluation of RBF SVM ***********")
print(classification_report(y_test,y_pred_rbf))
print("************* performance evaluation of SGD SVM ***********")
print(classification_report(y_test,y_pred_sgd))
print("************* performance evaluation of POLYNOMAIL SVM ***********")
print(classification_report(y_test,y_pred_poly))
print("************* performance evaluation of linear SVM ***********")
print(classification_report(y_test,y_pred_linear))
print("************* performance evaluation of random forest Classifier ***********")
print(classification_report(y_test,y_pred_forest))




************* performance evaluation of Naive bayes ***********
              precision    recall  f1-score   support

     negatif       0.86      0.97      0.91       746
      neutre       0.33      0.12      0.17        69
    positive       0.91      0.78      0.84       354

    accuracy                           0.86      1169
   macro avg       0.70      0.62      0.64      1169
weighted avg       0.84      0.86      0.85      1169

************* performance evaluation of MLP Classifier  ***********
              precision    recall  f1-score   support

     negatif       0.88      0.93      0.91       746
      neutre       0.26      0.10      0.15        69
    positive       0.85      0.84      0.85       354

    accuracy                           0.86      1169
   macro avg       0.66      0.63      0.63      1169
weighted avg       0.83      0.86      0.84      1169

************* performance evaluation of RBF SVM ***********
              precision    recall  f1-score   

In [19]:
pickle.dump(vectorizer,open('vectorizer.pickle','wb'))
pickle.dump(linear_svm,open('bestmodelfortunisiancomment.pickle','wb'))